presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeFilter.java

```diff
@@ -0,0 +1,98 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.analyzer;
+
+import com.facebook.presto.metadata.FunctionRegistry;
+import com.facebook.presto.sql.tree.DefaultExpressionTraversalVisitor;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.GroupingOperation;
+import com.facebook.presto.sql.tree.Node;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public final class ExpressionTreeFilter
+{
+    private ExpressionTreeFilter() {}
+
+    public static Function<Node, Optional<Expression>> groupingExtractionFunction = (node -> {
+        if (node instanceof GroupingOperation) {
+            return Optional.of((GroupingOperation) node);
+        }
+        return Optional.empty();
+    });
+
+    public static Function<Node, Optional<Expression>> windowExtractorFunction = (node -> {
+        if (node instanceof FunctionCall) {
+            FunctionCall functionCall = (FunctionCall) node;
+            if (functionCall.getWindow().isPresent()) {
+                return Optional.of(functionCall);
+            }
+        }
+        return Optional.empty();
+    });
+
+    public static Function<Node, Optional<Expression>> aggregateExtractorFunction(FunctionRegistry functionRegistry)
+    {
+        return ((node) -> {
+            if (node instanceof FunctionCall) {
+                FunctionCall functionCall = (FunctionCall) node;
+                if ((functionRegistry.isAggregationFunction(functionCall.getName()) || functionCall.getFilter().isPresent()) && !functionCall.getWindow().isPresent()) {
+                    return Optional.of(functionCall);
+                }
+            }
+            return Optional.empty();
+        });
+    }
+
+    public static List<Expression> extract(Function<Node, Optional<Expression>> predicate, List<? extends Node> nodes)
+    {
+        requireNonNull(predicate, "predicate is null");
+        requireNonNull(nodes, "nodes is null");
+
+        ImmutableList.Builder<Expression> extractedExpressions = ImmutableList.builder();
+        for (Node node : nodes) {
+            List<? extends Node> flattenedNodes = linearizeNodes(node);
+            extractedExpressions.addAll(flattenedNodes.stream()
+                    .map(predicate)
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .collect(toImmutableList()));
+        }
+
+        return extractedExpressions.build();
+    }
+
+    private static List<Node> linearizeNodes(Node node)
```

Comment:
See https://github.com/prestodb/presto/pull/6704#discussion_r106983581. Some of the inputs derive directly from `Node` and not `Expression`.